Allow running Linux System apps for foreign targets (#1603)
beeware/briefcase: docs/how-to/internal/x11passthrough.rst

======================
Docker X11 Passthrough
======================

The ability to run apps targeting other Linux distributions was added in
`beeware/briefcase#1603 <https://github.com/beeware/briefcase/pull/1603>`__.
This how-to explains the details of exposing the host environment to the Docker
container in which the app runs.

X Window System Background
--------------------------

> Nearly all modern Linux distributions use `Version 11
> <https://www.x.org/releases/X11R7.7/doc/xproto/x11protocol.html>`__ of the X
> Window System Protocol (X11) as the foundation for their graphical displays.
> Notably, though, `Wayland <https://wayland.freedesktop.org/>`__ is gaining
> traction as the successor to X11. Due to reasons of entrenchment, Wayland will
> likely continue providing their X11 compatibility layer to support X
> applications well in to the future.

Reviewer: ```suggestion
Linux distributions use either `X Windows <https://www.x.org/>`__(sometimes called X or
X11) or `Wayland <https://wayland.freedesktop.org/>`__ to manage their graphical
displays. X11 is the older of the two; Wayland maintains compatibility with the `X11
protocol <https://www.x.org/releases/X11R7.7/doc/xproto/x11protocol.html>`__ for
backwards compatibility.
```

Author: hmmm...."X Windows" sounds like msft has something to do with this? What about "X Window Manager"?

The X11 protocol operates in a client/server framework; any application that
wishes to display a window or receive user input will send and receive commands
with an X server.

X Configuration
~~~~~~~~~~~~~~~

The location of the X server is declared by the ``DISPLAY`` environment
variable, normally in the form ``HOSTNAME:DISPLAYNUMBER.SCREENNUMBER``. If
``HOSTNAME`` is absent, it is assumed to be the machine the client is running
on. The ``SCREENNUMBER`` is largely historical since all monitors are normally
collapsed together in to a single screen now. Therefore, an expected setting
for ``DISPLAY`` is ``:0`` for many installations.

While security is relatively weak for X11, there are basic facilities to
mitigate unauthenticated access. The ``XAUTHORITY`` environment variable can
specify a file path to an ``xauth`` database; if ``XAUTHORITY`` is not set,

> there will normally be a system default file path configured foe each user. The
> ``xauth`` database, itself, is protected by file system access controls and

Reviewer: ```suggestion
there will normally be a system default file path configured for each user. The
``xauth`` database itself is protected by file system access controls and
```

will contain "cookies" for individual displays that assigned by the X server
for clients to use to facilitate authentication.

> An example ``xauth`` database:

Reviewer: ```suggestion
An ``xauth`` database authorising a host named ``jupiter`` might look like:
```

.. code-block:: console

   $ xauth list
   jupiter/unix:  MIT-MAGIC-COOKIE-1  9e9a67185b1fdc0c46e00dc400559873
   #ffff#6a757069746572#:  MIT-MAGIC-COOKIE-1  9e9a67185b1fdc0c46e00dc400559873

Along with cookie-based authentication, it is also possible to add entities to

> an allowlist for a display. For instance, some distros are configured to allow

Reviewer: ```suggestion
an allowlist for a display. For instance, some distributions are configured to allow
```

any process owned by the logged in user access to the display. This is
configured by ``xhost``.

> Example ``xhost`` configuration:

Reviewer: ```suggestion
An ``xhost`` configuration authorising a user named ``brutus`` might look like:
```

.. code-block:: console

  $ xhost
  access control enabled, only authorized clients can connect

>   SI:localuser:russell

Reviewer: ```suggestion
  SI:localuser:brutus
```

X Operation
~~~~~~~~~~~

While authentication is normally enabled for X access, security is mostly
bolstered by the allowed methods for clients to connect. Most systems will only
open a UNIX socket to which clients should connect to send and receive
messages. By virtue of file system design, only users on the host machine will
have access to this socket. With the introduction of abstract sockets in Linux,
such a socket is also typically made available in tandem with the UNIX socket.
The advantages of an abstract socket is beyond this discussion, though.

Such a UNIX socket connection will be configured for each display for the
machine. As with many things in Linux, UNIX sockets are exposed in the machine
via a file in the ``/tmp/.X11-unix/`` directory. The socket files are named for
the number of the DISPLAY they are connected to; so, the file for Display 0 is
``X0``. Therefore, any display can be found at ``/tmp/.X11-unix/X#``.

Along with a UNIX socket connection, X servers can also listen on a TCP socket
on the machine's network interfaces. However, since a network connection can
easily be reached by other machines, listening on a TCP socket is normally

> disabled on most distros. That said, the X11 standard reserves port numbers

Reviewer: ```suggestion
disabled on most Linux distributions. That said, the X11 standard reserves port numbers
```

machine. Therefore, when Docker Desktop runs a container, it is running it
inside of a VM and not directly on the host system as Docker Engine does
(albeit in isolation via containerization).

As an outcome of the design of Docker Desktop, the behavior of interactions
between the host machine and Docker containers can be significantly different.
For instance, it is not possible to expose the host machine's network to a
container via ``--net host`` like you can with Docker Engine. While this does
change the exact network configuration that's exposed to the container in the
Docker Desktop VM, it is much different than Docker Engine and abstract sockets
on the host are not available to the container.

Along with not being possible to expose abstract sockets on the host to a
container running via Docker Desktop, it is also not possible to expose
arbitrary UNIX sockets either. Therefore, attempting to bind mount
``/tmp/.X11-unix/X0``, for instance, in to a Docker Desktop container will not
allow processes inside the container to successfully communicate with the
socket. (The Docker team has added support to pass specific sockets such as the

> socket Docker, itself, uses as well as the SSH agent socket; but exposing

Reviewer: ```suggestion
socket Docker itself uses, as well as the SSH agent socket; but exposing
```

container whether it is Docker Engine or Docker Desktop starting it.

In conclusion, we can add ``--add-host host.docker.internal:host-gateway`` to
the options to start a container and the host network interface will be
reachable at ``host.docker.internal``.

Exposing an X Display to a Container
------------------------------------

Given the knowledge of the operation of the Docker implementations, we finally
have the pieces to expose an X display to a container. Since it is not possible
to expose the display's socket directly to a container, a TCP proxy is
configured to pass X messages on the network from the container to the socket
on the host machine for the display.

TCP Proxy
~~~~~~~~~

The `socat <https://repo.or.cz/socat.git>`__ tool is a widely available program
to relay bi-directional data transfer between independent data channels. It
allows running a process on the host to listen on a network port and send any
received data to a socket connected to an X display on the other side.

Creating a TCP proxy for the X display effectively creates a spoofed X display.
The proxy is configured to listen on the TCP port for an unallocated display;

> the port number will be 6000 + the number of the display. Additionally, the

Reviewer: Is this a standard, a convention, or something specific to Briefcase?

Author: Defined in the X11 standard.

https://www.x.org/releases/X11R7.7/doc/man/man1/Xorg.1.xhtml#heading5

proxy is configured to listen on all network interfaces since identifying the
exact interface that will be available within the container is non-trivial.

The other side of the proxy is connected to the socket for the X display. The
socket, though, for the display may actually be another TCP socket; this will
be the case if the environment is currently configured for X11 forwarding over
SSH, as discussed below. In most cases, though, the socket will be the UNIX
socket for the display in the ``/tmp/.X11-unix/`` directory.

X Authentication for the Proxied X Display
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Authentication for X displays is managed in ``xauth`` database files. The
``xauth`` program allows for reading and writing the database. The database is
relatively simple mapping of display information to a cookie. When a client
wants to establish a connection for a display, it queries the database for the
display and receives the cookie back.

Since the proxy creates a spoofed display, a new ``xauth`` database needs to be
created for the spoofed display using the authentication afforded to the user
for the current display.

> Steps to create new database:

Reviewer: ```suggestion
To create new database, you need to:
```

 - Extract the cookie for the current display
 - Create a new database file
 - Add an entry for the spoofed display using the extracted cookie to the new
   database

>  - Rewrite the hostname of the entry that was just created to be "FamilyWild"

Reviewer: What's the significance of the FamilyWild name?

Author: It's the term the specs for `xauth` use.

https://gitlab.freedesktop.org/xorg/lib/libxau

> The Family is specified in two bytes to allow out-of-band values
(i.e. values not in the Protocol) to be used.  In particular,
two new values "FamilyLocal" and "FamilyWild" are defined.  FamilyLocal
refers to any connections using a non-network method of connection from the
local machine (Unix domain sockets, shared memory, loopback serial line).
In this case the host address is specified by the data returned from
gethostname() and better be unique in a collection of machines
which share NFS directories.  FamilyWild is currently used only
by xdm to communicate authorization data to the server.  It matches
any family/host address pair.
